### PR TITLE
Add perfect-numbers exercise

### PR DESCRIPTION
**Pull request by Akshat-mehrotra**

### CODE REVIEW

**Comment by Akshat-mehrotra** [comment]
How do I add it to the config file?

**Comment by Akshat-mehrotra** [comment]
I have made a generator for test files using LazyJSON. It works well on my fork but on the server side for some reason, I can't read an external JSON file. This is why my runtests.jl looks this ugly

**Comment by Akshat-mehrotra** [comment]
some of the tests take 3 to 4 seconds to run. The CLI quits too early !!

**Comment by SaschaMann** [comment]
> some of the tests take 3 to 4 seconds to run. The CLI quits too early !!

Could you elaborate? Unless they fail or finish, the tests will keep going for up to 6 hours.

**Comment by Akshat-mehrotra** [comment]
after checking the logs, it seems like the CI test quits because it doesn't receive any output from the test.jl file for about 10 mins ( I said 3 or 4 seconds in my earlier comment but it was about 13 to 14 mins). I am trying to implement a work around for that.

**Comment by SaschaMann** [comment]
Let me know when you're done pushing changes and I can do a proper review.

One thing though: if you're generating the tests by using a script, you need to save its output to `runtests.jl`. The generator itself should not be exposed to it. You can place it in `generators/perfect-numbers.jl`

**Comment by Akshat-mehrotra** [comment]
this seems logical, I'll implement that

**Comment by Akshat-mehrotra** [comment]
How do I "save" the output of the generator. My generator is just reading data from the json and running tests then and there. How do I get its output and save it in another .jl file

**Comment by Akshat-mehrotra** [comment]
Finally!!

File: exercises/perfect-numbers/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+Determine if a number is perfect, abundant, or deficient based on
```
Comment (SaschaMann) [inline exercises/perfect-numbers/README.md]: The readme needs to be generated from the templates with the configlet. You can find more information in the `CONTRIBUTING.md` and in [Generating a README](https://github.com/exercism/docs/blob/master/language-tracks/exercises/anatomy/readmes.md#generating-a-readme)

File: exercises/perfect-numbers/generator/canonical-data.json [hunk 2]
```diff
@@ -0,0 +1,134 @@
+{
```
Comment (SaschaMann) [inline exercises/perfect-numbers/generator/canonical-data.json]: This information should either be downloaded from GitHub or loaded from a local copy of the problem-specs repo. Generators exist to make it easier to update test suites when there are changes to the shared test data, inserting a fixed version here prevents that.

File: exercises/perfect-numbers/generator/copy and pase code from here.txt [hunk 3]
```diff
@@ -0,0 +1,39 @@
+@testset "Smallest perfect number is classified correctly" begin
```
Comment (SaschaMann) [inline exercises/perfect-numbers/generator/copy and pase code from here.txt]: I think this is a temporary file that was accidentally committed?

File: exercises/perfect-numbers/generator/generator.jl [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# canonical version = 1.3.0
```
Comment (SaschaMann) [inline exercises/perfect-numbers/generator/generator.jl]: See above about the fixed version.

This file should be in [`generators/`](https://github.com/exercism/julia/tree/master/generators) instead.

File: exercises/perfect-numbers/generator/generator.jl [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+using Test
+
+using LazyJSON
```
Comment (SaschaMann) [inline exercises/perfect-numbers/generator/generator.jl]: If you introduce new generator dependencies, please add them to the environment in [`generators/`](https://github.com/exercism/julia/tree/master/generators). Otherwise it will be harder to reproduce a working environment later on when running them again.

File: exercises/perfect-numbers/generator/generator.jl [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+j = LazyJSON.value(String(read("generator\\canonical-data.json")))
```
Comment (SaschaMann) [inline exercises/perfect-numbers/generator/generator.jl]: Please use `joinpath` to join paths in a cross-platform way. This would fail on non-Windows machines.

File: exercises/perfect-numbers/generator/generator.jl [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+for i in j["cases"]
+    for j in i["cases"]
+        des = String(j["description"])
+        inp = j["input"]["number"]
+
+        i["description"] == "Invalid inputs" ? exped = j["expected"]["error"] : exped = j["expected"]
+        global CODE *= """
+        @testset \"$des\" begin
+           @test classify($inp) == \"$exped\"
+        end
```
Comment (SaschaMann) [inline exercises/perfect-numbers/generator/generator.jl]: Really minor, but perhaps add a new line here to make it more consistent with the style/formatting of the other test suites.

File: exercises/perfect-numbers/metadata.yml [hunk 8]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (SaschaMann) [inline exercises/perfect-numbers/metadata.yml]: This file shouldn't be added to this repo. It's only used to generate the README but it'll be taken from the problem-spec repo when doing that.

File: exercises/perfect-numbers/runtests.jl [hunk 9]
```diff
@@ -0,0 +1,42 @@
+using Test
+include("example.jl")
```
Comment (SaschaMann) [inline exercises/perfect-numbers/runtests.jl]:
```suggestion
include("perfect-numbers.jl")
```

The student doesn't have access to `example.jl` when solving the exercise, so this would fail. To run the tests locally, use `julia runtests.jl perfect-numbers` at the root of this repo. It will automatically load the example instead of the stub file.

File: exercises/perfect-numbers/runtests.jl [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+@testset "Smallest perfect number is classified correctly" begin
+   @test classify(6) == "perfect"
+end
+@testset "Medium perfect number is classified correctly" begin
+   @test classify(28) == "perfect"
+end
+@testset "Large perfect number is classified correctly" begin
+   @test classify(33550336) == "perfect"
+end
```
Comment (SaschaMann) [inline exercises/perfect-numbers/runtests.jl]: The canonical data groups these together. This should also be reflected in the Julia tests by using (nested) testsets.

File: exercises/perfect-numbers/runtests.jl [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+@testset "Edge case (no factors other than itself) is classified correctly" begin
+   @test classify(1) == "deficient"
+end
+@testset "Zero is rejected (not a natural number)" begin
+   @test classify(0) == "Classification is only possible for natural numbers."
+end
+@testset "Negative integer is rejected (not a natural number)" begin
+   @test classify(-1) == "Classification is only possible for natural numbers."
+end
```
Comment (SaschaMann) [inline exercises/perfect-numbers/runtests.jl]: In the `canonical-data.json`, objects with an `error` key mean that the function should throw an error when called.

```json
                {
                    "description": "Negative integer is rejected (not a natural number)",
                    "property": "classify",
                    "input": {
                        "number": -1
                    },
                    "expected": {
                        "error": "Classification is only possible for natural numbers."
                    }
                }
```

translates to 

```julia
@testset "Zero is rejected (not a natural number)" begin
    @test_throws DomainError classify(0)
end
```

I chose `DomainError` because it fits this particular case, but `@test_throws` can be used with any kind of error.

File: config.json [hunk 12]
```diff
@@ -611,6 +611,24 @@
                 "data_structures",
                 "algorithms"
             ]
+        },
+        {
+            "uuid": "4fc89443-1c5c-492a-850f-7990869caae6",
+            "slug": "perfect-numbers",
+            "core": true,
```
Comment (SaschaMann) [inline config.json]:
```suggestion
            "core": false,
```

I'll have to think about where this fits into the progression, let's add it as floating exercise for now.

File: exercises/perfect-numbers/generator/generator.jl [hunk 13]
```diff
@@ -1,23 +1,54 @@
...
+jsonfileurl = "https://raw.githubusercontent.com/exercism/problem-specifications/master/exercises/perfect-numbers/canonical-data.json"
+jsonfilename = "canonical-data.json"
+jsonfile = download(jsonfileurl, joinpath(dir,jsonfilename))
+outputfile = "parse code.txt"
+
+
+f = LazyJSON.parse(read(joinpath(dir,jsonfile)))
+
+CODE = """
+# canonical version = $(j["version"])
```
Comment (SaschaMann) [inline exercises/perfect-numbers/generator/generator.jl]:
```suggestion
# canonical data version $(j["version"])
```

The dashboard parses this line so it needs to have a specific format

**Comment by Akshat-mehrotra** [comment]
all your suggestions made logical sense. idk why I didn't think of them before

**Comment by SaschaMann** [comment]
That's what code review is for :D It's very easy to miss small things like taht while writing code. Could you also run the generator again or fix the version comment in the `runtests.jl` file manually?

**Comment by Akshat-mehrotra** [comment]
sorry for the late reply, I got busy with some other assignment. Anyway I did it

**Comment by SaschaMann** [comment]
Sorry, I couldn't take a look yet. Why does the PR add a labeler config?

**Comment by logankilpatrick** [comment]
@SaschaMann just for fun! Was part of an un-related GCI task so the student chose to add a PR labeler to this repo. We can always remove it if you don't want it.

**Comment by SaschaMann** [comment]
Can we move the labeler to a different PR please, to keep a cleaner commit history?

File: exercises/perfect-numbers/generator/parse code.txt [hunk 14]
```diff
@@ -0,0 +1,69 @@
+# canonical data version 1.1.0
```
Comment (SaschaMann) [inline exercises/perfect-numbers/generator/parse code.txt]: this file should be removed again.

File: exercises/perfect-numbers/generator/canonical-data.json (hunk 2 above)
Comment (SaschaMann) [inline exercises/perfect-numbers/generator/canonical-data.json]: this too

File: exercises/perfect-numbers/generator/generator.jl [hunk 15]
```diff
@@ -0,0 +1,54 @@
+using Test
```
Comment (SaschaMann) [inline exercises/perfect-numbers/generator/generator.jl]: this should be moved to the `generators/perfect-numbers.jl`, not `exercises/perfect-numbers/generator/generator.jl`. Sorry I didn't notice that earlier.

**Comment by logankilpatrick** [comment]
Bump @Akshat-mehrotra

**Comment by Akshat-mehrotra** [comment]
Sorry for the long silence. My exams start in 2 days.

File: .github/labeler.yml [hunk 16]
```diff
@@ -0,0 +1,6 @@
+docs:
```
Comment (logankilpatrick) [inline .github/labeler.yml]:
```suggestion
docs: test 
```

File: exercises/perfect-numbers/README.md [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+  - 6 is a perfect number because (1 + 2 + 3) = 6
+  - 28 is a perfect number because (1 + 2 + 4 + 7 + 14) = 28
+- **Abundant**: aliquot sum > number
+  - 12 is an abundant number because (1 + 2 + 3 + 4 + 6) = 16
+  - 24 is an abundant number because (1 + 2 + 3 + 4 + 6 + 8 + 12) = 36
+- **Deficient**: aliquot sum < number
+  - 8 is a deficient number because (1 + 2 + 4) = 7
+  - Prime numbers are deficient
+
+Implement a way to determine whether a given number is **perfect**. Depending on your language track, you may also need to implement a way to determine whether a given number is **abundant** or **deficient**.
```
Comment (cmcaine) [inline exercises/perfect-numbers/README.md:18]:
```suggestion
Implement a way to determine whether a given number is **perfect**, **abundant** or **deficient**.
```

This readme is for the Julia track

File: exercises/perfect-numbers/example.jl [hunk 18]
```diff
@@ -0,0 +1,21 @@
+function classify(number)
+    factors = []
+    if number == 1
+        return "deficient"
```
Comment (cmcaine) [inline exercises/perfect-numbers/example.jl:4]: Maybe we should have this return symbols instead?

File: .github/labeler.yml [hunk 19]
```diff
@@ -0,0 +1,6 @@
...
+- "**/*.md"
+exercise:
+- "exercises/**"
+config change:
+- "config.json"
```
Comment (cmcaine) [inline .github/labeler.yml:6]: This should be in a different PR.

You can add a commit deleting it and make a new PR including it :)

**Comment by cmcaine** [comment]
I'll fix these remaining issues in a new PR :)
